Change MachXO2 IO encoding a bit

egorxe opened the pull request:
Removed CLAMP bits from IO BASE_TYPE and removed some default values. Addresses https://github.com/YosysHQ/prjtrellis/issues/254.

egorxe committed:
Change MachXO2 IO encoding a bit. Remove CLAMP bits from IO BASE_TYPE and remove some default values. Fixes incorrect IO type encoding in some cases.

egorxe committed:
Fix IO encoding for MachXO2 LRC1PIC2 tile and for XO3 & XO3D IO tiles

egorxe committed:
Fix OSCH/J.NOM_FREQ enums for values ending with 0

mmicko commented:
There are `.config_enum PIOC.DRIVE 8`  and `.config_enum PIOD.DRIVE 8`  (in  PIC_B0 for XO3/XO3D) so I assume PIOC and PIOD are not taken care of.
Understood that default value for CLAMP should not exists, but there are `.config_enum PIOA.CLAMP OFF` (for B and C and D as well)) but only in XO2 files and that PIC_BO of XO3/XO3D. Suggest you do text search in generated bit files to confirm there are no mentions of patterns you are trying to remove.

egorxe committed:
Fix some more IO encoding for MachXO3/3D

egorxe committed:
Remove MachXO2/3/3D default values for IO PULLMODE

egorxe commented:
> There are `.config_enum PIOC.DRIVE 8` and `.config_enum PIOD.DRIVE 8` (in PIC_B0 for XO3/XO3D) so I assume PIOC and PIOD are not taken care of. 

XO3&XO3D 051-pio_attrs fuzzers were failing on DIFFRESISTOR fuzzing which is done only on "B" tiles and I've missed it, sorry. The error message from Diamond was "Differential comp pad is not placed on a true pad of the true/complementary pair", I've changed fuzzers a bit to only try DIFFRESISTOR for A and C PIOs. Now all the tiles for all three families should be coherent.

>Understood that default value for CLAMP should not exists

Sorry for confusion here cause I've done it differently for XO2 and XO3. Actually I think that default values should be removed only for enums which have bits intersecting with IO BASE_TYPE and cant be easily disentangled from it, so they may mess with IO type. These are DRIVE, OPENDRAIN, PULLMODE and DIFFRESISTOR. I've removed defaults for them and brought defaults for CLAMP back on all families.

egorxe committed:
Small fix for MachXO2 NOM_FREQ enum

mmicko commented:
Are you removing default values manually, or this is result of fuzzer ?

egorxe commented:
Generally these are results of fuzzers, which I update here https://github.com/YosysHQ/prjtrellis/pull/255 . But I've only ran fuzzers 051, 052, 059, 069 and 102 stitching everything together with scripts and sometimes manually to save time, cause I have only a single Diamond license and fuzzing in one thread takes a lot of time. I could rerun all existing fuzzers for all three families if you think it's necessary to verify that the result will be the same. But it will take several days or more if there are more problems with failing fuzzers like DIFFRESISTOR for XO3.

egorxe committed:
Fix more default values for MachXO3/3D

egorxe commented:
@mmicko It took some time, but I've rerunned all fuzzers for all 3 families and now I'm sure that the PIO-related values in database are totally consistent with fuzzing result. The only other differences to the original database that I got after my fuzzer run are some minor discrepancies in fixed_conns for some tiles. I'm not sure where they do come from (different Diamond version may be?) but they have nothing to do with my changes here. Will you please consider merging it now?

mmicko commented:
@egorxe Thanks. Merged in both repos, and I have updated link to latest database at the end. Thanks for you hard work on this, I know it was not easy at all.
